test/framework/exec/docker.go

```diff
@@ -0,0 +1,324 @@
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package exec
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
+)
+
+const (
+	httpProxy  = "HTTP_PROXY"
+	httpsProxy = "HTTPS_PROXY"
+	noProxy    = "NO_PROXY"
+)
+
+var dockerClient *client.Client
+
+// getDockerClient returns a new client connection for interacting with the Docker engine.
+func getDockerClient() (*client.Client, error) {
+	if dockerClient == nil {
+		var err error
+		dockerClient, err = client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+		if err != nil {
+			return nil, err
+		}
+	}
+	return dockerClient, nil
+}
+
+// SaveImage saves a Docker image to the file specified by dest.
+func SaveDockerImage(ctx context.Context, image, dest string) error {
+	docker, err := getDockerClient()
+	if err != nil {
+		return err
+	}
+
+	reader, err := docker.ImageSave(ctx, []string{image})
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+
+	tar, err := os.Create(dest)
+	if err != nil {
+		return err
+	}
+	defer tar.Close()
+
+	_, err = io.Copy(tar, reader)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// PullDockerImage triggers the Docker engine to pull an image if it is not
+// already present.
+func PullDockerImage(ctx context.Context, image string) error {
+	docker, err := getDockerClient()
+	if err != nil {
+		return err
+	}
+
+	pullResp, err := docker.ImagePull(ctx, image, types.ImagePullOptions{})
+	if err != nil {
+		return err
+	}
+	defer pullResp.Close()
+
+	// Clients must read the ImagePull response to EOF to complete the pull
+	// operation or errors can occur.
+	_, err = io.ReadAll(pullResp)
+	if err != nil {
+		return nil
+	}
+
+	return nil
+}
+
+// GetLoadBalancerHostPort looks up the host port bound for the load balanced API.
+func GetLoadBalancerHostPort(ctx context.Context, name string) (string, error) {
+	docker, err := getDockerClient()
+	if err != nil {
+		return "", err
+	}
+
+	// Get details about the container
+	containerInfo, err := docker.ContainerInspect(ctx, name)
+	if err != nil {
+		return "", err
+	}
+
+	// Loop through the container port bindings and return the first HostPort
+	for port, bindings := range containerInfo.NetworkSettings.Ports {
+		if port == "6443/tcp" {
```

Comment:
Is this always correct?

Reply:
Yes, this is hard coded for the load balancer port. If you look at the existing code, this is just moving that same logic from CLI filtering to checking programmatically.